app-backend/src/main/kotlin/org/example/appbackend/config/JwtTokenFilter.kt

```diff
@@ -0,0 +1,103 @@
+import java.util.Date
+import java.util.*
+import org.slf4j.LoggerFactory
+
+class JwtTokenFilter(
+        @Value("\${spring.jwt.secret}")
+        private val jwtSecret: String,
+        private val userService: UserService
+) : OncePerRequestFilter() {
+
+    private var logger = LoggerFactory.getLogger(JwtTokenFilter::class.java)
+
+    @Throws(ServletException::class, IOException::class)
+    override fun doFilterInternal(
+            request: HttpServletRequest,
+            response: HttpServletResponse,
+            filterChain: FilterChain
+    ) {
+        val authorizationHeader = request.getHeader("Authorization")
+
+        var userId: Int? = null
+        var jwtToken: String?
+
+        if (authorizationHeader != null && authorizationHeader.startsWith("Bearer ")) {
+            jwtToken = authorizationHeader.substring(7)
+            logger.info("/JwtTokenFilter: jwtToken= $jwtToken")
+            try {
+                val session = request.session
+                val token = session.getAttribute("token") as? String
+                logger.info("/JwtTokenFilter: session token = $token")
+                if (token != jwtToken) {
+                    throw IllegalArgumentException("Token not found.")
+                }
+                userId = extractUserId(jwtToken)
+                logger.info("/JwtTokenFilter: userId = $userId")
+            } catch (e: IllegalArgumentException) {
+                // Log the error
+                logger.error("Invalid JWT token: ${e.message}")
+                // Return HTTP unauthorized response
+                response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid JWT token")
+                return
+            }
+        }
+
+        if (userId != null && SecurityContextHolder.getContext().authentication == null) {
+            val userDetails = userService.getUserById(userId)
+            logger.info("/JwtTokenFilter: username= ${userDetails.username}")
+            if (userDetails != null) {
+                val authenticationToken = UsernamePasswordAuthenticationToken(
+                        userDetails, null, null
+                )
+                authenticationToken.details = WebAuthenticationDetailsSource().buildDetails(request)
+                SecurityContextHolder.getContext().authentication = authenticationToken
+                logger.info("/JwtTokenFilter: end.")
+            }
+        }
+        filterChain.doFilter(request, response)
+    }
+
+    private fun extractUserId(token: String): Int {
+        val keyBytes = jwtSecret.toByteArray(Charsets.UTF_8) // Correct encoding to Byte array
+        val key: Key = Keys.hmacShaKeyFor(keyBytes)
+
+        val claimsJws: Jws<Claims> = try {
+            Jwts.parserBuilder()
+                .setSigningKey(key)
+                .build()
+                .parseClaimsJws(token)
+        } catch (e: JwtException) {
+            throw IllegalArgumentException("Invalid JWT token: ${e.message}")
```

Comment:
This is the better way: `throw IllegalArgumentException("Invalid JWT token: ${e.message}", e)`